examples/timeseries_classification_from_scratch.qmd

```diff
@@ -49,109 +50,118 @@ Here we visualize one timeseries example for each class in the dataset.
+classification problems, see [Bagnall et al.
+(2016)](https://link.springer.com/article/10.1007/s10618-016-0483-9).
 
 In order to use `sparse_categorical_crossentropy`, we will have to count
 the number of classes beforehand.
 
 ```{r}
-num_classes <- length(unique(y_train[[1]]))
+(num_classes <- length(unique(y_train)))
 ```
 
-Now we shuffle the training set because we will be using the `validation_split` option
-later when training.
+Now we shuffle the training set because we will be using the
+`validation_split` option later when training.
 
 ```{r}
-train_ind <- sample(nrow(x_train))
-x_train <- x_train[train_ind,]
-y_train <- y_train[train_ind,]
+shuffle_ind <- sample(nrow(x_train))
+x_train <- x_train[shuffle_ind, ]
+y_train <- y_train[shuffle_ind, ]
 ```
 
-Standardize the labels to positive integers.
-The expected labels will then be 0 and 1.
+Standardize the labels to positive integers. The expected labels will
+then be 0 and 1.
 
 ```{r}
-y_train[y_train == -1, ] <- 0
-y_test[y_test == -1, ] <- 0
+y_train[y_train == -1] <- 0
+y_test[y_test == -1] <- 0
 ```
 
-Note that the timeseries data used here are univariate, meaning we only have one channel
-per timeseries example.
-We will therefore transform the timeseries into a multivariate one with one channel
-using a simple reshaping via numpy.
-This will allow us to construct a model that is easily applicable to multivariate time
-series.
+Note that the timeseries data used here are univariate, meaning we only
+have one channel per timeseries example. We will therefore transform the
+timeseries into a multivariate one with one channel using a simple
+reshaping. This will allow us to construct a model that is easily
+applicable to multivariate time series.
 
 ```{r}
-x_train <- as.matrix(x_train)
+# add channel dim of size 1
 dim(x_train) <- c(dim(x_train), 1)
-x_test <- as.matrix(x_test)
 dim(x_test) <- c(dim(x_test), 1)
 ```
 
 ## Build a model
 
 We build a Fully Convolutional Neural Network originally proposed in
-[this paper](https://arxiv.org/abs/1611.06455).
-The implementation is based on the TF 2 version provided
-[here](https://github.com/hfawaz/dl-4-tsc/).
-The following hyperparameters (kernel_size, filters, the usage of BatchNorm) were found
-via random search using [KerasTuner](https://github.com/keras-team/keras-tuner).
+[this paper](https://arxiv.org/abs/1611.06455). The implementation is
+based on the TF 2 version provided
+[here](https://github.com/hfawaz/dl-4-tsc/). The following
+hyperparameters (kernel_size, filters, the usage of BatchNorm) were
+found via random search using
+[KerasTuner](https://github.com/keras-team/keras-tuner).
 
 ```{r}
-input_shape <- dim(x_train)[-1]
+input_shape <- dim(x_train)[-1] # drop batch dim
 input_layer <- layer_input(input_shape)
 
 output_layer <- input_layer %>%
+
   # First convolutional layer
-  layer_conv_1d(filters = 64, kernel_size = 3, padding = "same") %>% 
+  layer_conv_1d(64, 3, padding = "same") %>%
```

Comment:
This was intentional. I remember I had to explicitly pass these names arguments due to some errors in newer TF version. I’d double check though.

Reply:
the `layer_*` functions do two things:
1. Instantiate a layer instance
2. Compose the layer instance with the first argument passed to `layer_*()`, return the result of the composition. 

"composition" here can mean calling `.add()` on a sequential model and returning the sequential model, or `__call__()`ing the layer instance and returning the resultant tensor.

If you want the `layer_*()` function to skip step 2 and just return the layer instance, the
first argument `object` needs to be missing or NULL. One way to do that is to supply all the other arguments as named arguments, thus avoiding any positional matching and ensuring that the first argument to the function is missing. 

Note that if you do `inputs %>% layer_conv_1d(64)`, `64` is the 2nd argument to the call, _not_ the first.